# 导入应用的库

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from __future__ import division


In [2]:
churn_df = pd.read_csv("churn.csv",encoding='gbk')
#c查看全部特征的字段
data_ls = churn_df.columns.tolist()
display(data_ls)

['客户姓名',
 '顾客编号',
 '区号',
 '电话',
 '活动1',
 '活动2',
 '小产品销售',
 '羊初乳体验',
 '越桔体验',
 '黑曜石手链',
 '维生素B',
 '体验时间',
 '饺子宴',
 'N',
 'O',
 'P',
 'Q',
 '主产品销售',
 'S',
 '会员等级',
 '是否流失']

In [3]:
churn_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4746 entries, 0 to 4745
Data columns (total 21 columns):
客户姓名     4746 non-null object
顾客编号     4746 non-null int64
区号       4746 non-null int64
电话       4746 non-null object
活动1      4746 non-null object
活动2      4746 non-null object
小产品销售    4746 non-null int64
羊初乳体验    4746 non-null float64
越桔体验     4746 non-null int64
黑曜石手链    4746 non-null float64
维生素B     4746 non-null float64
体验时间     4746 non-null int64
饺子宴      4746 non-null float64
N        4746 non-null float64
O        4746 non-null int64
P        4746 non-null float64
Q        4746 non-null float64
主产品销售    4746 non-null int64
S        4746 non-null float64
会员等级     4746 non-null int64
是否流失     4746 non-null object
dtypes: float64(8), int64(8), object(5)
memory usage: 778.7+ KB


In [4]:
to_show = data_ls[:2] + data_ls[-6:]
churn_df[to_show].head(6)

客户姓名  顾客编号      P     Q  主产品销售     S  会员等级    是否流失
0   KS   128  11.01  10.0      3  2.70     1  False.
1   OH   107  11.45  13.7      3  3.70     1  False.
2   NJ   137   7.32  12.2      5  3.29     0  False.
3   OH    84   8.86   6.6      7  1.78     2  False.
4   OK    75   8.41  10.1      3  2.73     3  False.
5   AL   118   9.18   6.3      6  1.70     0  False.

In [5]:
churn_result = churn_df['是否流失']
y = np.where(churn_result == 'True.',1,0)
display(y)

array([0, 0, 0, ..., 0, 0, 0])

In [6]:
# 我们不需要这些列
to_drop = ['客户姓名','顾客编号','电话','是否流失','N','O','P','Q','S','区号']
churn_feat_space = churn_df.drop(to_drop,axis=1)
display(churn_feat_space.head(10))


活动1  活动2  小产品销售  羊初乳体验  越桔体验  黑曜石手链   维生素B  体验时间    饺子宴  主产品销售  会员等级
0   no  yes     25  265.1   110  45.07  197.4    99  16.78      3     1
1   no  yes     26  161.6   123  27.47  195.5   103  16.62      3     1
2   no   no      0  243.4   114  41.38  121.2   110  10.30      5     0
3  yes   no      0  299.4    71  50.90   61.9    88   5.26      7     2
4  yes   no      0  166.7   113  28.34  148.3   122  12.61      3     3
5  yes   no      0  223.4    98  37.98  220.6   101  18.75      6     0
6   no  yes     24  218.2    88  37.09  348.5   108  29.62      7     3
7  yes   no      0  157.0    79  26.69  103.1    94   8.76      6     0
8   no   no      0  184.5    97  31.37  351.6    80  29.89      4     1
9  yes  yes     37  258.6    84  43.96  222.0   111  18.87      5     0

In [7]:
#“是”/“否”必须转换为布尔值
#NumPy将这些值从布尔值转换为1。和0。晚些时候
yes_no_cols = ["活动1","活动2"]
churn_feat_space[yes_no_cols] = churn_feat_space[yes_no_cols] == 'yes'
display(churn_feat_space[yes_no_cols].head(10))


活动1    活动2
0  False   True
1  False   True
2  False  False
3   True  False
4   True  False
5   True  False
6  False   True
7   True  False
8  False  False
9   True   True

In [8]:
# 提取功能供将来使用
features = churn_feat_space.columns

X = churn_feat_space.as_matrix().astype(np.float)
display(X)

D:\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  after removing the cwd from sys.path.


array([[ 0.  ,  1.  , 25.  , ..., 16.78,  3.  ,  1.  ],
       [ 0.  ,  1.  , 26.  , ..., 16.62,  3.  ,  1.  ],
       [ 0.  ,  0.  ,  0.  , ..., 10.3 ,  5.  ,  0.  ],
       ...,
       [ 0.  ,  0.  ,  0.  , ..., 18.03,  5.  ,  1.  ],
       [ 0.  ,  0.  ,  0.  , ..., 16.35,  6.  ,  1.  ],
       [ 0.  ,  1.  , 24.  , ..., 19.82,  4.  ,  1.  ]])

In [9]:
# 标准化处理 
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X = scaler.fit_transform(X)

display("Feature space holds %d observations and %d features" % X.shape) 
display("Unique target labels:", np.unique(y))  
display(X[0])  
display(len(y[y == 0]))

'Feature space holds 4746 observations and 11 features'

'Unique target labels:'

array([0, 1])

array([-0.32760674,  1.6101965 ,  1.22868967,  1.56179583,  0.47750718,
        1.56206732, -0.06319954, -0.05703268, -0.06301713, -0.60079374,
       -0.42790886])

4081

# 进行标准化处理

In [10]:
from sklearn.cross_validation import KFold

def run_cv(X,y,clf_class,**kwargs):
    # 切分数据集
    kf = KFold(len(y),n_folds=5,shuffle=True)
    y_pred = y.copy()

    # 遍历
    for train_index, test_index in kf:
        X_train, X_test = X[train_index], X[test_index]
        y_train = y[train_index]
        # 初始化一个分类器
        clf = clf_class(**kwargs)
        clf.fit(X_train,y_train)
        y_pred[test_index] = clf.predict(X_test)
    return y_pred

D:\Anaconda3\lib\site-packages\sklearn\cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [11]:
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier as RF
from sklearn.neighbors import KNeighborsClassifier as KNN

def accuracy(y_true,y_pred):
    # 真（1） 假（0）
    return np.mean(y_true == y_pred)

display("支持向量机:") 
display("%.3f" % accuracy(y, run_cv(X,y,SVC)))
display( "随机深林:")
display("%.3f" % accuracy(y, run_cv(X,y,RF)))
display("k近邻:") 
display("%.3f" % accuracy(y, run_cv(X,y,KNN)))

'支持向量机:'

'0.925'

'随机深林:'

'0.961'

'k近邻:'

'0.907'

In [22]:
def run_prob_cv(X, y, clf_class, **kwargs):
    kf = KFold(len(y), n_folds=7, shuffle=True)
    y_prob = np.zeros((len(y),2))
    for train_index, test_index in kf:
        X_train, X_test = X[train_index], X[test_index]
        y_train = y[train_index]
        clf = clf_class(**kwargs)
        clf.fit(X_train,y_train)
        # Predict probabilities, not classes
        y_prob[test_index] = clf.predict_proba(X_test)
    return y_prob

# 转换为概率值

In [23]:
import warnings
warnings.filterwarnings('ignore')

# 使用10个估计值，这样预测都是0.1的倍数
pred_prob = run_prob_cv(X, y, RF, n_estimators=10)
display( pred_prob[0])

array([0.8, 0.2])

In [24]:
pred_churn = pred_prob[:,1]
is_churn = y == 1

#预测概率被分配给观测的次数
counts = pd.value_counts(pred_churn)
display( counts)



0.0    2987
0.1     716
0.2     256
1.0     190
0.9     135
0.3     118
0.8      78
0.4      76
0.6      74
0.7      68
0.5      48
dtype: int64

In [25]:
# 计算正确的概率
true_prob = {}
for prob in counts.index:
    true_prob[prob] = np.mean(is_churn[pred_churn == prob])
    true_prob = pd.Series(true_prob)

# pandas-fu
counts = pd.concat([counts,true_prob], axis=1).reset_index()
counts.columns = ['pred_prob', 'count', 'true_prob']
counts

pred_prob  count  true_prob
0         0.0   2987   0.011717
1         0.1    716   0.032123
2         0.2    256   0.097656
3         1.0    190   1.000000
4         0.9    135   0.962963
5         0.3    118   0.152542
6         0.8     78   0.961538
7         0.4     76   0.328947
8         0.6     74   0.716216
9         0.7     68   0.941176
10        0.5     48   0.562500

# 项目结论 
# 根据数据集 以流失 未流失的客户  
# 当可能流失的概率为0.8时 有78人 而真实流失的概率为0.96
# 而 0.7时 真实流失的概率为0.94   也十分高
# 根据业务需求 选取需要的概率值
# 再根据（实际业务需求）一个用户可能带来的效益 大于 一个新客户的效益 
# 选定概率值后 0.7后 我们对 新样本中预测为70%流失概率的用户进行
# 发信息 打电话 用一些礼品去维护 和 挽回 这些用户 